[MRG+1] Enable using 2SD scaling in StandardScaler

parth1993 opened the pull request:
#### Reference Issues/PRs
Fixes #10864

#### What does this implement/fix? Explain your changes.
Enables using 2SD scaling in StandardScaler.

References:
Gelman, A. (2008). Scaling regression inputs by dividing by two standard deviations.
Statistics in medicine, 27(15), 2865-2873.

parth1993 committed:
Updated StandardScaler

twolodzko commented on sklearn/preprocessing/data.py:
```diff
@@ -508,9 +508,10 @@ class StandardScaler(BaseEstimator, TransformerMixin):
-    with_std : boolean, True by default
-        If True, scale the data to unit variance (or equivalently,
-        unit standard deviation).
+    with_std : boolean or int, 1 by default
```
Maybe: boolean, 1 or 2, True by default

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -508,9 +508,10 @@ class StandardScaler(BaseEstimator, TransformerMixin):
+    with_std : boolean or int, 1 by default
+        If 1 or 2, scale the data to unit variance (or equivalently,
+        unit standard deviation) using scaling factor of 1 or 2 standard 
+        deviation.
```
Add "See [1]_" and define the Gelman reference as 1

jnothman reviewed (commented):
This needs tests for the valid and invalid cases

jnothman commented:
Please use "Fixes #xxx" in the pr description so GitHub knows to automatically close the issue.

parth1993 commented:
@jnothman Ok. Thanks.

parth1993 committed:
Update data.py

parth1993 reviewed (commented):
Added Gelman reference. Please review.

lesteve commented:
You have some PEP8 violations, please invest some time to configure on-the-fly flake8 checks inside your editor of choice. This allows you to spot these kind of problems while editing.

parth1993 commented:
@lesteve. Thanks a lot I will follow this and will also learn.

parth1993 commented:
@lesteve will also look for PEP8 violations.

parth1993 committed:
Update data.py

parth1993 reviewed (commented):
Fixed PEP8 violations

parth1993 committed:
Update data.py

parth1993 reviewed (commented):
Fixed PEP8 violations

parth1993 committed:
Update data.py

jnothman reviewed (commented):
Please add a test in test_data.py

jnothman commented:
Please add an entry to the change log at `doc/whats_new/v0.20.rst`. Like the other entries there, please reference this pull request with `:issue:` and credit yourself (and other contributors if applicable) with `:user:`

parth1993 commented:
@jnothman Should I add it as class or Fix?

jnothman commented:
enhancement

On 29 Mar 2018 9:22 am, "Techno_pyparth" <notifications@github.com> wrote:

> @jnothman <https://github.com/jnothman> Should I add it as class or Fix?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10875#issuecomment-377057345>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6ys9UwTadORKAEqEbOL-FyNcIFkDks5tjA0IgaJpZM4S7tDD>
> .
>

parth1993 commented:
Ok thanks.

parth1993 commented:
@jnothman Sorry but I need help in writing test. Any thoughts on test cases?

parth1993 committed:
Update v0.20.rst

parth1993 committed:
Enable using 2SD scaling in StandardScaler (#10864)

jnothman reviewed (commented):
See other tests in sklearn/preprocessing/tests/test_data.py. you can use made up data, you just need to check that scale_ is correctly set.

parth1993 committed:
Added Gelman ref.

parth1993 committed:
Update v0.20.rst

parth1993 committed:
Resolved merge conflict.

parth1993 committed:
Added test cases for StandardScaler

parth1993 committed:
Fix flake8 compatibility

parth1993 commented:
@jnothman Please review this commit.

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -508,10 +508,11 @@ class StandardScaler(BaseEstimator, TransformerMixin):
-        unit standard deviation).
-
+    with_std : boolean, 1 or 2, True by default
+        If 1 or 2, scale the data to unit variance (or equivalently,
```
If 2, it's no longer "unit variance".

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -508,10 +508,11 @@ class StandardScaler(BaseEstimator, TransformerMixin):
+    with_std : boolean, 1 or 2, True by default
+        If 1 or 2, scale the data to unit variance (or equivalently,
+        unit standard deviation) using scaling factor of 1 or 2 standard 
+        deviation. See [1]_
```
the reference to [1] should explicitly about 2.

jnothman reviewed (commented):
I'm now wondering if we should use with_std=0.5 instead of 2. But certainly, the confusion between 0.5 and 2 is a reason for us to be raising an error if with_std is not {0/False, 1/True, 2}. Please handle this case.

parth1993 committed:
Added reference to [1] explicitly about 2.
Added error raise if `with_std` is not {0/False, 1/True, 2}.
Updated `with_std` documentation.

parth1993 committed:
Fix Flake8 compatibility

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -508,10 +508,12 @@ class StandardScaler(BaseEstimator, TransformerMixin):
-
+        unit standard deviation). 
+        Otherwise 2, scale the data to using scaling factor 2
+        standard deviation. See [1] for `with_std==2`.
```
Let's use = not == here

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -661,7 +667,11 @@ def partial_fit(self, X, y=None):
 
         if self.with_std:
-            self.scale_ = _handle_zeros_in_scale(np.sqrt(self.var_))
+            if self.with_std is not True and self.with_std is not 2:
```
How about `with_std not in (1, 2, True)`

parth1993 replied:
Sure, that's make sense.

jnothman reviewed (commented):
Please also test input validation

Please also make similar changes to the scale function

parth1993 committed:
Added extra parameter `with_std` value i.e. {2} for scale function.
Added reference to [1] explicitly about 2
Added error raise if `with_std` is not {0/False, 1/True, 2}.
Updated `with_std` documentation.
Tested input validation.

parth1993 commented:
@jnothman Please review the PR. I'm not able to understand the codecov/patch issue.

jnothman reviewed (commented):
Codecov is complaining that your changes to scale are not tested

lesteve commented:
Looking at [this](https://codecov.io/gh/scikit-learn/scikit-learn/compare/97523985b39ecde369d83352d7c3baf403b60a22...2f59b802281a38b945c26d73790bea79836ac39a/src/sklearn/preprocessing/data.py?before=sklearn/preprocessing/data.py), it looks like you are not testing the error cases when `with_std not in (1, 2, True)`. You should add a test with `assert_raise_regex`.

parth1993 committed:
Added test cases for invalid `with_std`

parth1993 committed:
Added test coverage

jnothman commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2376,3 +2376,33 @@ def test_power_transformer_lambda_zero():
+    X = rng.randint(0, 2, (100, 2))
+    scaler = StandardScaler(copy=True, with_mean=True, with_std=2)
+    X_scaled = scaler.fit(X).transform(X, copy=True)
+    assert_array_almost_equal(X_scaled.mean(), 0.0)
```
Since you're not comparing arrays, this is confusing. Please use `assert np.isclose`

parth1993 replied:
Okay, will do the same.

jnothman commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2376,3 +2376,33 @@ def test_power_transformer_lambda_zero():
+    assert_array_almost_equal(X_scaled.std(), 1.0)
+
+
+def test_scale_with_invalid_std():
```
I don't think this should say invalid

parth1993 replied:
Okay, will change it! :+1:

parth1993 committed:
Minor changes

parth1993 commented:
I think it should be noted.

parth1993 commented:
Reference of SD is not there. Should I include one?

jnothman commented:
it mentions unit variance

parth1993 commented:
Yes, I read it. Should I include reference of Gelman's argument in a new line?

jnothman commented:
sure. try and we'll review

parth1993 committed:
Enhanced StandardScaler

parth1993 committed:
PEP8 fixes

parth1993 committed:
Merge branch 'master' into feature

parth1993 committed:
Minor fixes

parth1993 committed:
Resolved merge conflict by incorporating both suggestions.

parth1993 committed:
PEP8 fixes

parth1993 committed:
Fixes

parth1993 commented:
@jnothman Please review.

adrinjalali commented on doc/modules/preprocessing.rst:
```diff
@@ -25,7 +25,9 @@ Standardization, or mean removal and variance scaling
-normally distributed data: Gaussian with **zero mean and unit variance**.
+normally distributed data: Gaussian with **zero mean and unit/factor two standard**
+**deviation** in reference to **Gelman, A. (2008). Scaling regression inputs by** 
+**dividing by two standard deviations. Statistics in medicine, 27(15), 2865-2873**.
```
I think you can add a reference bellow the document as

```
.. [1] Gelman, A. (2008). Scaling regression inputs by
dividing by two standard deviations. Statistics in medicine, 27(15), 2865-2873
```

and refer to it in the document.

adrinjalali commented on doc/whats_new/v0.20.rst:
```diff
@@ -591,9 +591,19 @@ Support for Python 3.3 has been officially dropped.
   and :user:`Hanmin Qin <qinhanmin2014>`.
 
+
+- :class:`preprocessing.PolynomialFeatures` now supports sparse input.
```
I think we'll include this in v0.21 and not 0.20 (since it's kind of a new feature), please move the entry to the corresponding file.

jnothman replied:
This is a merge error. Use `git checkout master doc/whats_new/v0.20.rst; git commit`

thomasjpfan replied:
This still needs to be addressed.

adrinjalali commented on sklearn/preprocessing/data.py:
```diff
@@ -97,9 +97,11 @@ def scale(X, axis=0, with_mean=True, with_std=True, copy=True):
-    with_std : boolean, True by default
-        If True, scale the data to unit variance (or equivalently,
-        unit standard deviation).
+    with_std : boolean, 1 or 2, True by default
```
Isn't this confusing a bit? As a user, I'd expect to be able to give any floating point value as the input, if the API's accepting numbers for the scaling factor. Maybe string literals `'one', 'two'` would be more appropriate? Otherwise, I don't see why we'd limit the user to these two values.

parth1993 replied:
Right we shouldn't limit user to these two values. Instead it should be `'int'`

adrinjalali replied:
or `float`

adrinjalali replied:
but I'd like to know what @jnothman thinks about this one.

adrinjalali commented on sklearn/preprocessing/data.py:
```diff
@@ -148,15 +154,22 @@ def scale(X, axis=0, with_mean=True, with_std=True, copy=True):
             raise ValueError("Can only scale sparse matrix on axis=0, "
                              " got axis=%d" % axis)
         if with_std:
+            if with_std not in (1, 2, True):
```
you can move the validation part before the `if` and avoid code duplication.

adrinjalali commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2303,6 +2303,35 @@ def test_power_transformer_lambda_zero():
+    # Test for invalid cases
+    cases = [-10, -1, 3, 10]
+    for std_ in cases:
+        assert_raises_regex(ValueError, r'Invalid value for `with_std`:'
```
please use `with pytest.raises(...)` construct instead.

parth1993 committed:
Minor changes

jnothman commented:
Hmmmm. I was happy to restrict the accepted values, and extend it if users
complain, but maybe it doesn't matter. But I'm now not certain about the
issue... This scaling is proposed to assist in coefficient interpretation.
We have elsewhere rejected implementing sample standard deviation (ddof)
which similarly just adjusts the standard deviation by a constant
multiplicative factor (assuming no missing values).

parth1993 committed:
Changes

parth1993 committed:
Merge branch 'master' of https://github.com/parth1993/scikit-learn into my-feature

parth1993 committed:
Fix

parth1993 committed:
pytest construct

parth1993 commented:
Please review the changes.

adrinjalali commented:
please remove the accidentally added binary files.

parth1993 committed:
Delete binary file

parth1993 committed:
Delete binary file

parth1993 commented:
> please remove the accidentally added binary files.

Done

parth1993 commented:
Please review the changes

parth1993 commented:
Please review the changes

parth1993 commented:
@jnothman @adrinjalali please review the PR

parth1993 commented:
@jnothman @adrinjalali @twolodzko Will this PR be merged?

thomasjpfan commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2303,6 +2303,35 @@ def test_power_transformer_lambda_zero():
+
+
+@pytest.mark.parametrize('with_std', [-10, -1, 3, 10])
+def test_scale_with_std(with_std):
```
Please split this tests into two test:

1. `test_scale_with_std()` to just test `scale`
2. `test_scale_with_std_error(with_std)` to test the `ValueError`. In `pytest.raises` uses the `match` keyword to assert the text of the error message.

parth1993 replied:
Done

thomasjpfan commented on sklearn/preprocessing/data.py:
```diff
@@ -731,7 +749,11 @@ def partial_fit(self, X, y=None):
-            self.scale_ = _handle_zeros_in_scale(np.sqrt(self.var_))
+            if self.with_std not in (1, 2, True):
+                raise ValueError("Invalid value for 'with_std': %s" %
+                                 str(self.with_std))
```
`"%s"` converts python objects use `str()`:

```suggestion
                                 self.with_std)
```

thomasjpfan commented on sklearn/preprocessing/data.py:
```diff
@@ -148,15 +154,21 @@ def scale(X, axis=0, with_mean=True, with_std=True, copy=True):
             raise ValueError("Can only scale sparse matrix on axis=0, "
                              " got axis=%d" % axis)
         if with_std:
```
Is this check still needed if we have `if with_std not in (...)`?

parth1993 replied:
@thomasjpfan Yes, so we can restrict the user to (1, 2, True) values.

parth1993 replied:
Yes, it's an additional check.

thomasjpfan replied:
We can raise an error earlier by including the following check right after `X = check_array(...)`

```py
if with_std not in (1, 2, True, False):
    raise ValueError("Invalid value for `with_std`: {}".format(
                     str(with_std)))
```

And remove the two checks we have now.

thomasjpfan commented on sklearn/preprocessing/data.py:
```diff
@@ -148,15 +154,21 @@ def scale(X, axis=0, with_mean=True, with_std=True, copy=True):
         X = np.asarray(X)
         if with_mean:
             mean_ = np.nanmean(X, axis)
         if with_std:
```
Is this check still needed if we have `if with_std not in (...)`?

parth1993 replied:
@thomasjpfan Yes, so we can restrict the user to (1, 2, True) values.

parth1993 replied:
Yes, it's an additional check.

parth1993 committed:
Update sklearn/preprocessing/data.py

Co-Authored-By: Thomas J Fan <thomasjpfan@gmail.com>

Commit:
Changes

parth1993 committed:
Changes

parth1993 committed:
update

parth1993 commented:
Can someone help me with circleci : doc and doc-min-dependencies are failing.

rth commented:
> Can someone help me with circleci : doc and doc-min-dependencies are failing.

You should merge master in -- that will fix it.

parth1993 committed:
Merge branch 'master' into my-feature

parth1993 commented:
Please review PR.

parth1993 committed:
Merge remote-tracking branch 'upstream/master' into my-feature

parth1993 committed:
Check Fixed

parth1993 committed:
Flake8 Fixes

thomasjpfan commented on sklearn/preprocessing/data.py:
```diff
@@ -136,10 +138,18 @@ def scale(X, axis=0, with_mean=True, with_std=True, copy=True):
     X = check_array(X, accept_sparse='csc', copy=copy, ensure_2d=False,
                     estimator='the scale function', dtype=FLOAT_DTYPES,
                     force_all_finite='allow-nan')
+    if with_std:
```
`if with_std not in (1, 2, True, False):`

parth1993 replied:
E    ValueError: Invalid value for `with_std`: False

parth1993 replied:
Done

parth1993 committed:
Changes

thomasjpfan commented:
From Gelman's paper, using `with_std=2` is beneficial because:

> Here we propose dividing each numeric variable by two times its standard deviation, so that the generic comparison is with inputs equal to the mean ±1 standard deviation. The resulting coefficients are then directly comparable for untransformed binary predictors.

In a follow up PR, we can include an example on when `with_std=2` is useful.

parth1993 commented:
Sure.

parth1993 commented:
Sure, then what should be the next step?

On Tue, 25 Jun 2019, 23:47 Thomas J Fan, <notifications@github.com> wrote:

> From Gelman's paper, using with_std=2 is beneficial because:
>
> Here we propose dividing each numeric variable by two times its standard
> deviation, so that the generic comparison is with inputs equal to the mean
> ±1 standard deviation. The resulting coefficients are then directly
> comparable for untransformed binary predictors.
>
> In a follow up PR, we can include an example on when with_std=2 is useful.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10875?email_source=notifications&email_token=AD6GUUQFAMEIFCECGDNN3NTP4JOM5A5CNFSM4EXO2DB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYREB5A#issuecomment-505561332>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD6GUUQUDAWJKMPL6ETEQTLP4JOM5ANCNFSM4EXO2DBQ>
> .
>

thomasjpfan commented:
@parth1993 May you comment on an example where using `with_std=2` is useful? A code example using a sample dataset would work.

amueller commented:
@thomasjpfan so did you want an example in a follow-up PR or this one?

thomasjpfan commented:
@amueller I would prefer new features to come along with an example of a use case.

@parth1993 Are you able to add an example to go with this PR?

amueller commented:
@thomasjpfan I agree but two comments above you said something else so I was confused ;)

parth1993 commented:
Yes, I would be able to do that by this weekend.

On Wed, 7 Aug 2019, 21:33 Thomas J Fan, <notifications@github.com> wrote:

> @amueller <https://github.com/amueller> I would prefer new features to
> come along with an example of a use case.
>
> @parth1993 <https://github.com/parth1993> Are you able to add an example
> to go with this PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10875?email_source=notifications&email_token=AD6GUURSO4APOAMOZN5EGCLQDLUBBA5CNFSM4EXO2DB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3Y2A2I#issuecomment-519151721>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD6GUUVHEP7CNEK3AOO3PTTQDLUBBANCNFSM4EXO2DBQ>
> .
>

parth1993 commented:
[StandardScaler-use-case.ipynb.zip](https://github.com/scikit-learn/scikit-learn/files/3501247/StandardScaler-use-case.ipynb.zip)

This is a use case where we can see the effects of parameter `with_std` on regression coefficients.

thomasjpfan commented:
Here is the notebook for reviewers:

https://nbviewer.jupyter.org/gist/thomasjpfan/b371b406b0b702b181e4101d87a5d1b4

I see that the coefficients has doubled. What can a user do with this new information? Gelman's paper argues that this is more interpretable.

I would be happy with an example in this PR that recreates his argument. This way when a user sees this `with_std=2` option, they can see an example on how to use it and why it is useful

(This means adding an `plot_*` in `examples/linear_model/` that uses `with_std=2`)

parth1993 commented:
@thomasjpfan yes an example in this PR that recreates his argument would work, but the data that Gelman used is not available. Any suggestions over this?

thomasjpfan commented:
There could be a dataset on https://www.openml.org that you can use to show the same arguments.

parth1993 commented:
Here is the notebook link:

https://nbviewer.jupyter.org/gist/parth1993/faa0d36a41b2fbe8d4476cab5c5183c3

I was able to produce an example of Gelman's argument.

thomasjpfan commented:
In your example, what does `with_std=2` provide that `with_std=1` does not?

parth1993 commented:
It's more a recommendation rescaling as a default option an improvement upon the usual approach of including variables in whatever way they are coded in the data file—so that the magnitudes of coefficients can be directly compared as a matter of routine statistical practice.  
Moreover, Gelman suggests ` the general practice of scaling numeric inputs by dividing
by two standard deviations, which allows the coefficients to be interpreted in the same way as
with binary inputs.`

thomasjpfan commented:
Imagine writing this as an example and you want to convince someone to use `with_std=2`. How does seeing a coefficient that is double make it more interpretable?

glemaitre commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2307,6 +2307,41 @@ def test_power_transformer_lambda_zero():
+    assert np.isclose(X_scaled.std(), 1.0)
+
+
+def test_scale_with_std():
```
Could you make a similar parametrization

glemaitre commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2307,6 +2307,41 @@ def test_power_transformer_lambda_zero():
+    # Test for invalid cases
+    with pytest.raises(ValueError) as excinfo:
+        scale(X, with_std=with_std)
+    assert excinfo.type is ValueError
```
This is checked by the `with` statement no need for that.

glemaitre commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2307,6 +2307,41 @@ def test_power_transformer_lambda_zero():
+    X = rng.randint(0, 2, (100, 2))
+
+    # Test for invalid cases
+    with pytest.raises(ValueError) as excinfo:
```
```suggestion
    with pytest.raises(ValueError, match='Invalid value for `with_value`'):
```

glemaitre commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2307,6 +2307,41 @@ def test_power_transformer_lambda_zero():
+
+    # Test for invalid cases
+    with pytest.raises(ValueError) as excinfo:
+        scale(X, with_std=with_std)
```
you should check the same error for `scaler.fit_transform`

glemaitre commented on doc/modules/preprocessing.rst:
```diff
@@ -25,7 +25,15 @@ Standardization, or mean removal and variance scaling
+Reference
+---------
+
+Gelman, A. (2008). Scaling regression inputs by 
```
```suggestion
    .. [Gelman2008] Gelman, A. (2008). Scaling regression inputs by 
```

glemaitre commented on doc/modules/preprocessing.rst:
```diff
@@ -25,7 +25,15 @@ Standardization, or mean removal and variance scaling
 badly if the individual features do not more or less look like standard
-normally distributed data: Gaussian with **zero mean and unit variance**.
+normally distributed data: Gaussian with **zero mean and unit/factor two standard**
+**deviation** in reference to **Gelman, A. (2008). Scaling regression inputs by** 
```
```suggestion
**deviation** [Gelman2008]_. 
```

glemaitre commented on doc/modules/preprocessing.rst:
```diff
@@ -25,7 +25,15 @@ Standardization, or mean removal and variance scaling
-normally distributed data: Gaussian with **zero mean and unit variance**.
+normally distributed data: Gaussian with **zero mean and unit/factor two standard**
+**deviation** in reference to **Gelman, A. (2008). Scaling regression inputs by** 
+**dividing by two standard deviations. Statistics in medicine, 27(15), 2865-2873**.
```
```suggestion
```

glemaitre commented on doc/modules/preprocessing.rst:
```diff
@@ -61,7 +69,10 @@ operation on a single array-like dataset::
         >>> np.set_printoptions(suppress=True)
 
-Scaled data has zero mean and unit variance::
+Scaled data has zero mean and unit/factor 2 standard deviation in
```
```suggestion
Scaled data has zero mean and unit or factor 2 standard deviation [Gelman2008]_.
```

glemaitre commented:
> Imagine writing this as an example and you want to convince someone to use with_std=2. How does seeing a coefficient that is double make it more interpretable?

Uhm reading a bit more the paper, it seems that you can start to compare binary variable (e.g. one-hot encoded) with a scaled variable. Te value taken by the one-hot encoded variable will be twice more than standard-scaled inputs. From the article, it would be:

```
A binary variable with equal probabilities has a mean 0.5 and standard deviation of 0.5. The usual standardized predictor (scaled by one standard deviation) then takes on the values±1, and a 1-unit difference on this transformed scale corresponds to a difference of 0.5 on the original variable (for example, a comparison betweenx=0.25 and 0.75), which cannot be directly interpreted
```

glemaitre commented:
So an example would be to take titanic and directly interpret the coefficients of a linear model considering age and sex for instance.

thomasjpfan commented:
> So an example would be to take titanic and directly interpret the coefficients of a linear model considering age and sex for instance.

I was concerned this was the only use case. This scaling would only be reasonable for one hot encoding for a binary feature. If we try to extend this to 5 categories, the scaling would need to go to 5?

glemaitre commented:
So it makes me think about the usefulness of this feature.

In practice, it would be equivalent of encoded categorical data with a one-hot encoder and then standardize all data (encoded and numeric). It might be good to introduce an example in the inspection to show how to do that and make some discussion but I don't think that having the parameter is good actually. Could be YAGNI

thomasjpfan commented:
For reference, here is what their R `standardize` implementation (from the paper) says:

> Numeric variables that take on more than two values are each rescaled to have a mean of 0 and asd of 0.5; Binary variables are rescaled to have a mean of 0 and a difference of 1 between their twocategories; Non-numeric variables that take on more than two values are unchanged; Variables thattake on only one value are unchanged

https://cran.r-project.org/web/packages/arm/arm.pdf Page 46

jnothman commented:
Yeah I'm quite ambivalent about this now

adrinjalali commented:
I'd be happy with the example.

parth1993 committed:
Update sklearn/preprocessing/data.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

parth1993 committed:
Update sklearn/preprocessing/data.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

parth1993 committed:
Update doc/modules/preprocessing.rst

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

parth1993 committed:
Update doc/modules/preprocessing.rst

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

parth1993 committed:
Update sklearn/preprocessing/data.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

parth1993 committed:
Update sklearn/preprocessing/data.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

parth1993 committed:
Update sklearn/preprocessing/data.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

parth1993 committed:
Update doc/modules/preprocessing.rst

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

parth1993 committed:
Update doc/modules/preprocessing.rst

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

amueller commented:
tangentially related: it would be nice if we could make ``RobustScaler`` behave like ``StandardScaler``. If we make this change to ``StandardScaler`` we might also want to add it to ``RobustScaler`` which might make things a bit more complicated? (also see #10140)

>  Indeed, we can already get similar results by plugging the StandardScaler in the "right

Can you please explain that @glemaitre?

glemaitre commented:
Using a `FunctionTransformer`, you can easily go from:

```python
numeric_features = ['age', 'fare']
numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())])

categorical_features = ['embarked', 'sex', 'pclass']
categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))])

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)])

# Append classifier to preprocessing pipeline.
# Now we have a full prediction pipeline.
clf = Pipeline(steps=[('preprocessor', preprocessor),
                      ('classifier', LogisticRegression())])

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

_ = clf.fit(X_train, y_train)
```

to

```python
numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler()),
    ('normalize', FunctionTransformer(func=lambda x: x / 2))])
```

Regarding the `StandardScaler`, I was thinking about this but this is wrong.

```python
numeric_features = ['age', 'fare']
numeric_transformer = SimpleImputer(strategy='median')

categorical_features = ['embarked', 'sex', 'pclass']
categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))])

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)])

# Append classifier to preprocessing pipeline.
# Now we have a full prediction pipeline.
clf = Pipeline(steps=[('preprocessor', preprocessor),
                      ('scaler', StandardScaler()),
                      ('classifier', LogisticRegression())])
```

But I am wrong because we want to always divide by 2 and not take the mean.

Alvaro-Kothe commented:
I intend to finish this pull request on #31268.

lorentzenchr commented on doc/modules/preprocessing.rst:
```diff
@@ -26,7 +25,14 @@ Standardization, or mean removal and variance scaling
 machine learning estimators** implemented in scikit-learn; they might behave
 badly if the individual features do not more or less look like standard
-normally distributed data: Gaussian with **zero mean and unit variance**.
+normally distributed data: Gaussian with **zero mean and unit/factor two standard**
+**deviation** [Gelman2008]_. 
```
```suggestion
normally distributed data: Gaussian with **zero mean and unit variance**.
For linear models [Gelman2008]_ advocates to rescale numerical feature by two instead of one standard deviations and leave binary indicators untouched, 0 or 1, see parameter `with_std`.
```

lorentzenchr commented on doc/modules/preprocessing.rst:
```diff
@@ -26,7 +25,14 @@ Standardization, or mean removal and variance scaling
+Reference
+---------
+
+    .. [Gelman2008] Gelman, A. (2008). Scaling regression inputs by 
+dividing by two standard deviations. Statistics in medicine, 27(15), 2865-2873
```
```suggestion
    .. [Gelman2008] Gelman, A. (2008).
       :doi:`Scaling regression inputs by dividing by two standard deviations.
       <10.1002/SIM.3107>`. Statistics in medicine, 27(15), 2865-2873
```

lorentzenchr commented on doc/modules/preprocessing.rst:
```diff
@@ -73,7 +68,10 @@ dataset::
         >>> np.set_printoptions(suppress=True)
 
-Scaled data has zero mean and unit variance::
+Scaled data has zero mean and unit or factor 2 standard deviation [Gelman2008]_.
+refernce to **Gelman, A. (2008). Scaling regression inputs by** 
+**dividing by two standard deviations. Statistics in medicine, 27(15)
+, 2865-2873**.::
```
```suggestion
The scaled data `X_scaled` has zero mean and unit standard deviation::
```

lorentzenchr commented on doc/whats_new/v0.21.rst:
```diff
@@ -945,6 +862,10 @@ Support for Python 3.4 and below has been officially dropped.
   in the dense case. Also added a new parameter ``order`` which controls output
   order for further speed performances. :pr:`12251` by `Tom Dupre la Tour`_.
 
+- |Feature| :class:`preprocessing.StandardScaler` can now scale using factor of 2
+  standard deviation as described in Gelman (2008). :issue:`10864` by 
+  :user:`Parth Sharma <parth1993>` and :user:`Tymoteusz Wołodźko <twolodzko>`
+
```
The new way of a whatsnew entry is to add a new file under `/doc/whats_new/upcoming_changes/sklearn.preprocessing`, see the Readme in the upcoming_changes folder.

lorentzenchr commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2278,26 +2307,57 @@ def test_power_transformer_lambda_zero():
     assert_array_almost_equal(pt.inverse_transform(X_trans), X)
 
 
+def test_standard_scaler_with_std():
```
```suggestion
def test_standard_scaler_with_std():
    """Test that ..."""
```
fill in a short description for the ...

lorentzenchr commented on sklearn/preprocessing/data.py:
```diff
@@ -0,0 +1,3032 @@
+        If True, center the data before scaling.
+
+    with_std : boolean, 1 or 2, True by default
+        If 1 or True, scale the data to unit variance (or equivalently,
+        unit standard deviation). 
+        Otherwise 2, scale the data to using scaling factor 2
+        standard deviation. See [1] for `with_std=2`.
```
```suggestion
        If 1 or True, scale the data to unit standard deviation (or equivalently,
        unit variance). 
        Otherwise 2, scale the data by dividing by two standard deviations, see [1]_.
```

lorentzenchr commented on sklearn/preprocessing/data.py:
```diff
@@ -0,0 +1,3032 @@
+        memory.
+
+    with_std : boolean, 1 or 2, True by default
+        - If `1` or `True`, scale the data to unit variance (or equivalently,
+        unit standard deviation);
+        - otherwise `2`, scale the data to using scaling factor 2
+        standard deviation. See [1] for `with_std=2`
```
Exact docstring as in `scale`.

lorentzenchr reviewed (commented):
This would be a nice little addition.

lorentzenchr commented:
BTW, I don't agree with the assessment in #31268 and the closing of this PR. I still think, this would be a valuable addition. Think of a preprocessing pipeline of linear models that uses OneHotEncoding for categoricals and StandardScaler for the rest. This also changes the impact of the penalty on the features (numerical will be more penalized compared to categoricals with this new scaling). This is in line with adding Bayesian priors to coefficients as done in http://dx.doi.org/10.1214/08-AOAS191.

Note: We now have an example on interpretation of linear model coefficients, see https://scikit-learn.org/stable/auto_examples/inspection/plot_linear_model_coefficient_interpretation.html.
